Trigger the onChange handler on the input component if specified

nielsvandoorn opened the pull request:
Hi @zackify,

Thank you for this great package. This pull request adds the possibility to have an onChange handler on the Input component. In the current version, onChange handlers are not invoked.

The onChange from the child component properties will be invoked if the method is passed as a property. I've also added a test for this.

Feel free to contact me if you have any questions.

Thanks,
Niels

audiolion commented:
@nielsvandoorn something must be wrong if onChange handlers aren't invoked, the existing code does invoke onChange handlers.

Right below where you inserted code we call `component.validateOnBlurOrChange` which accepts a function as the third argument

https://github.com/navjobs/validify/blob/c772546f670a4dfae1f4f13560878c22a2655b8f/src/input.js#L22-L27

That function is the `component.onChange` method and is called as the first part of the `validateOnBlurOrChange` function:

https://github.com/navjobs/validify/blob/c772546f670a4dfae1f4f13560878c22a2655b8f/src/base.js#L38-L39

Your PR would be duplicating the `onChange` event it seems

nielsvandoorn commented:
Could you try running the last test in tests/input.js that i've added in my branch with the original code? The test will fail with the error that the function was called 0 times.

I've also added a console.log statement at the ```component.validateOnBlurOrChange``` method, which will result in undefined in every single test case.

I have the exact same setup as in your code example, but still onChange handlers are not triggered. I would really appreciate if you could look into this. The test that i've added in my branch should help with that.

zackify commented:
```js
export const Input = ({ error,customOnChange, onChange, ...props }) => (
  <div>
    <p>{error}</p>
    <input {...props} onChange={e => {
       customOnChange(e)
       onChange(e)
     }} />
  </div>
);```

Is this what you want to do? Your own onChange as well as the form library itself?

nielsvandoorn commented:
Yes, I want to trigger custom functions on change (because I have a form that has dynamic fields which will show/hide based on selected values.

This would be the example in my case:
```
<InputField
    type="text"
    id="title"
    name="title"
    placeholder="Title"
    customOnChange={() => console.log('change', event)}
/>

export const InputField = ({ error, customOnChange, onChange, ...props }) => (
  <div style={{ width: '100%' }}>
    <input
      {...props}
      onChange={e => {
        customOnChange(e)
        onChange(e)
      }}
    />
    <Error>{error}</Error>
  </div>
)
```

But this code will error with ```customOnChange is not a function```

zackify commented:
Add a little if statement and this should work for you :)

```js

<InputField
    type="text"
    id="title"
    name="title"
    placeholder="Title"
    customOnChange={() => console.log('change', event)}
/>

export const InputField = ({ error, customOnChange, onChange, ...props }) => (
  <div style={{ width: '100%' }}>
    <input
      {...props}
      onChange={e => {
        if(customOnChange) customOnChange(e)
        onChange(e)
      }}
    />
    <Error>{error}</Error>
  </div>
)
```

zackify commented:
Oh also, you can just trigger something inside onValues as well. That would be a good way to do something any time there is a change.

nielsvandoorn commented:
Sure, but that still leaves me with the question why the onChange handler is not being called. 
I've tried this test on your current code:

```
test('Input onChange is triggered when passing onChange as prop', () => {
  const onChange = jest.fn();
  const wrapper = mount(
    <Form rules={{ Awesome: 'required' }}>
      <Input name="Awesome" onChange={() => onChange()} />
      <div className="submit" submit />
    </Form>
  );

  wrapper
    .find('input')
    .simulate('change', { target: { name: 'Awesome', value: 'Hello World' } });

  expect(onChange).toHaveBeenCalledTimes(1)
})
```

As you will see, this test will fail. If you will test again on my PR it will pass.
You could also extra check by passing a console.log statement in the onChange handler. It will not log anything to the console while testing but will with my changes.

zackify commented:
So the reason is that these inputs get an `onChange` `onBlur` and `onKeyUp` passed to them by the form component. 

Those three things are managed for you and you shouldn't need to pass your own onChange to any inputs.

If you wanted to do something when a field is changed you could check that inside the `onValues`, or call the onChange like I mentioned in the example. I hope this makes sense!

zackify commented:
What doesn't work in the code example? If it's not logging event thats because you had: 

`customOnChange={() => console.log('change', event)}`

needs to be `customOnChange={(event) => console.log('change', event)}`

This should completely work and if its not then we have a big problem, I can make a CRA doing this if you need.


This library was made so you never have to worry about onChange or other events that you would normally add to every input element repetitively.

nielsvandoorn commented:
Ofcourse I have the event as prop in the function, otherwise it would not work. But this example that you provided above is not logging anything to the console.

nielsvandoorn commented:
The whole problem is still that the onChange method on the Input is never evoked.

audiolion commented:
@nielsvandoorn hey man, sorry for giving bad info, I forgot that validify is passing its own `onChange` that is called.

As @zackify mentioned, if you want to call your own `onChange` then use the snippet he pasted above. The `onChange` is hijacked with the intent that it makes it easy for the developer to add form validation.

nielsvandoorn commented:
I understand why the package is hijacking the onChange handler, but with my
change is still gives the developer the ability to use it while also having
the onChange functionality in the package. The snippet simply just doesn't
work, because that onChange is hijacked by the package
Op do 23 aug. 2018 om 21:44 schreef Ryan Castner <notifications@github.com>

> @nielsvandoorn <https://github.com/nielsvandoorn> hey man, sorry for
> giving bad info, I forgot that validify is passing its own onChange that
> is called.
>
> As @zackify <https://github.com/zackify> mentioned, if you want to call
> your own onChange then use the snippet he pasted above. The onChange is
> hijacked with the intent that it makes it easy for the developer to add
> form validation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/navjobs/validify/pull/28#issuecomment-415547365>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5EUARvSuHMkWeMvrU1DXmvpx9bJ3vyks5uTwYbgaJpZM4WJsUT>
> .
>

nielsvandoorn commented:
I'm heading to bed now. Please take a brief look at my PR at https://github.com/nielsvandoorn/validify
I've also added a branch ```not-working``` to show an example (last test in tests/input.js) of how the code is not working

audiolion commented:
@nielsvandoorn you might be missing the fact that you need to provide `customOnChange` prop and call that from a modified input file. Can you check out the PR #29 and see if that makes sense? Attempt at updating the docs to cover this use-case

nielsvandoorn commented:
I did exactly that, and did not forget the customOnChange prop, but it
still doesn't work.
Op do 23 aug. 2018 om 21:53 schreef Ryan Castner <notifications@github.com>

> @nielsvandoorn <https://github.com/nielsvandoorn> you might be missing
> the fact that you need to provide customOnChange prop and call that from
> a modified input file. Can you check out the PR #29
> <https://github.com/navjobs/validify/pull/29> and see if that makes
> sense? Attempt at updating the docs to cover this use-case
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/navjobs/validify/pull/28#issuecomment-415550281>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5EUJ9DSIDg1y8iSvFPhwhWZucjJyGgks5uTwhLgaJpZM4WJsUT>
> .
>

audiolion commented:
@nielsvandoorn 

Your test should be something like:

```js
const Input = ({ error, myCustomOnChange, onChange, ...props }) => (
  <div>
    {error ? <p className="error">{error}</p> : null}
    <input 
      onChange={e => {
        if (myCustomOnChange) myCustomOnChange(e);
        onChange(e);
      }
      {...props}
    />
  </div>
);

test('Input onChange is triggered when passing onChange as prop', () => {
  const customOnChange = jest.fn();
  const wrapper = mount(
    <Form rules={{ Awesome: 'required' }}>
      <Input name="Awesome" myCustomOnChange={() => customOnChange()} />
      <div className="submit" submit />
    </Form>
  );

  wrapper
    .find('input')
    .simulate('change', { target: { name: 'Awesome', value: 'Hello World' } });

  expect(customOnChange).toHaveBeenCalledTimes(1)
})

nielsvandoorn commented:
@audiolion Hi, this works. But I'd still prefer the real onChange handler. The custom onChange method you added to the <input> element is kind of the same code as in my PR. Anyhow, I'm able to get the result I want. 

@zackify Feel free to close or merge the PR, the changes in the readme in #29 will be helpful to a lot of people.

audiolion commented:
@nielsvandoorn I think the only issue with the PR is that if, by habit, people place an onChange handler on there they will be unknowingly duplicating onChange calls not realizing the library already handles it. By leaving it this way with a custom on change we are making the user be explicit about the behavior they want. I can see where you are coming from though
